internal/graph/lookupsubjects.go

```diff
@@ -188,16 +189,142 @@ func (cl *ConcurrentLookupSubjects) lookupViaComputed(
 	parentRequest ValidatedLookupSubjectsRequest,
 	parentStream dispatch.LookupSubjectsStream,
-	ttu *core.TupleToUserset,
+	ttu *core.FunctionedTupleToUserset,
 ) error {
 	ds := datastoremw.MustFromContext(ctx).SnapshotReader(parentRequest.Revision)
 	it, err := ds.QueryRelationships(ctx, datastore.RelationshipsFilter{
 		OptionalResourceType:     parentRequest.ResourceRelation.Namespace,
-		OptionalResourceRelation: ttu.Tupleset.Relation,
+		OptionalResourceRelation: ttu.GetTupleset().GetRelation(),
+		OptionalResourceIds:      parentRequest.ResourceIds,
+	})
+	if err != nil {
+		return err
+	}
+	defer it.Close()
+
+	// TODO(jschorr): Find a more efficient means of doing this, if any.
+	collectingStreamsByIncomingResourceID := make(map[string][]*dispatch.CollectingDispatchStream[*v1.DispatchLookupSubjectsResponse])
+
+	cancelCtx, checkCancel := context.WithCancel(ctx)
+	defer checkCancel()
+
+	g, subCtx := errgroup.WithContext(cancelCtx)
+	g.SetLimit(int(cl.concurrencyLimit))
+
+	// For each found tuple, dispatch a lookup subjects request and collect its results.
+	// We need to intersect between *all* the found subjects for each resource ID.
+	var ttuCaveat *core.CaveatExpression
+
+	for tpl := it.Next(); tpl != nil; tpl = it.Next() {
+		if it.Err() != nil {
+			return it.Err()
+		}
+
+		// If the relationship has a caveat, add it to the overall TTU caveat. Since this is an intersection
+		// of *all* branches, the caveat will be applied to all found subjects, so this is a safe approach.
+		if tpl.Caveat != nil {
+			ttuCaveat = caveatAnd(ttuCaveat, wrapCaveat(tpl.Caveat))
+		}
+
+		if err := namespace.CheckNamespaceAndRelation(ctx, tpl.Subject.Namespace, ttu.GetComputedUserset().Relation, false, ds); err != nil {
+			if !errors.As(err, &namespace.ErrRelationNotFound{}) {
+				return err
+			}
+
+			continue
+		}
+
+		resourceID := tpl.ResourceAndRelation.ObjectId
+		stream := dispatch.NewCollectingDispatchStream[*v1.DispatchLookupSubjectsResponse](subCtx)
+		if _, ok := collectingStreamsByIncomingResourceID[resourceID]; !ok {
+			collectingStreamsByIncomingResourceID[resourceID] = []*dispatch.CollectingDispatchStream[*v1.DispatchLookupSubjectsResponse]{}
+		}
+
+		collectingStreamsByIncomingResourceID[resourceID] = append(collectingStreamsByIncomingResourceID[resourceID], stream)
+
+		tpl := tpl
+		g.Go(func() error {
+			return cl.d.DispatchLookupSubjects(&v1.DispatchLookupSubjectsRequest{
+				ResourceRelation: &core.RelationReference{
+					Namespace: tpl.Subject.Namespace,
+					Relation:  ttu.GetComputedUserset().Relation,
+				},
+				ResourceIds:     []string{tpl.Subject.ObjectId},
+				SubjectRelation: parentRequest.SubjectRelation,
+				Metadata: &v1.ResolverMeta{
+					AtRevision:     parentRequest.Revision.String(),
+					DepthRemaining: parentRequest.Metadata.DepthRemaining - 1,
+				},
+			}, stream)
+		})
+	}
+	it.Close()
+
+	// Wait for all dispatched operations to complete.
+	if err := g.Wait(); err != nil {
```

Comment:
You shouldn't need to wait here - since this is an `AND` operation, you can just add in the sets as you get them (it's associative and commutative). That should keep memory lower too.